### DISCUSSION TITLE
How can I stream to a MemoryStream, error: (StoreProhibited)

### DISCUSSION THREAD

**Discussion by ksaye** [pschatzmann/arduino-audio-tools#669] [Q&A] [answered]
In the code below, I get the following error.  My ultimate goal is to write to a memory object so the next step could be to read the memory object so I can upload the wav file to a webstite.

      #include "AudioTools.h"
      
      #define MicPIN_WS GPIO_NUM_33
      #define MicPIN_SCK GPIO_NUM_32
      #define MicPIN_SD GPIO_NUM_36
      
      I2SStream MicStream;                                                        // I2S stream for the INMP441 I2S Microphone
      MemoryStream ms(512, RAM);
      uint8_t *waveStream[512];
      StreamCopy micToMem(ms, MicStream);
      
      void setup()
      {
          Serial.begin(115200);
          AudioLogger::instance().begin(Serial, AudioLogger::Debug);
          WiFi.begin("ixxxxx", "Fyyyyyyy");
          while (WiFi.status() != WL_CONNECTED) {
              delay(500);
              Serial.print(".");
          }
      
          I2SConfig config2 = MicStream.defaultConfig(RX_MODE);
          config2.port_no = 1;
          config2.channels = 1;
          config2.sample_rate = 16000;
          config2.bits_per_sample = 16;
          config2.pin_ws = MicPIN_WS;       //WS
          config2.pin_bck = MicPIN_SCK;     //SCK
          config2.pin_data = MicPIN_SD;     //SD
          config2.use_apll = false;
          config2.auto_clear = true;
          config2.buffer_size = 512;
          config2.buffer_count = 16;
          MicStream.begin(config2);
          micToMem.setCheckAvailableForWrite(false);
      }
      
      void loop()
      {
          if (MicStream.available() and ms.availableForWrite()){
              micToMem.copy();
          }
      }

Here is my error:

      ..................[D] I2SStream.h : 53 - bool audio_tools::I2SStream::begin(audio_tools::I2SConfig)
      [D] I2SESP32.h : 44 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfig)
      [D] I2SESP32.h : 140 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfig, int, int)
      [I] I2SConfig.h : 104 - rx/tx mode: RX_MODE
      [I] I2SConfig.h : 105 - port_no: 1
      [I] I2SConfig.h : 106 - is_master: Master
      [I] I2SConfig.h : 107 - sample rate: 16000
      [I] I2SConfig.h : 108 - bits per sample: 16
      [I] I2SConfig.h : 109 - number of channels: 1
      [I] I2SConfig.h : 110 - i2s_format: I2S_STD_FORMAT
      [I] I2SConfig.h : 112 - auto_clear:1
      [I] I2SConfig.h : 119 - buffer_count:16
      [I] I2SConfig.h : 120 - buffer_size:512
      [I] I2SConfig.h : 124 - pin_bck: 32
      [I] I2SConfig.h : 126 - pin_ws: 33
      [I] I2SConfig.h : 128 - pin_data: 36
      [D] I2SESP32.h : 172 - i2s_driver_install
      [D] I2SESP32.h : 189 - i2s_set_pin
      [D] I2SESP32.h : 200 - i2s_zero_dma_buffer
      [D] I2SESP32.h : 204 - begin - started
      [D] AudioCopy.h : 92 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
      Guru Meditation Error: Core  1 panic'ed (StoreProhibited). Exception was unhandled.
      Core 1 register dump:
      PC      : 0x401420c8  PS      : 0x00060c30  A0      : 0x801420fc  A1      : 0x3ffb1ef0  
      A2      : 0x00000000  A3      : 0x00000000  A4      : 0x3ffb1f00  A5      : 0x00000100  
      A6      : 0x00000000  A7      : 0x00000002  A8      : 0x3ffc16cc  A9      : 0x00000000  
      A10     : 0x00000200  A11     : 0x00000002  A12     : 0x00000000  A13     : 0x00000100  
      A14     : 0x00000000  A15     : 0x00000002  SAR     : 0x00000020  EXCCAUSE: 0x0000001d  
      EXCVADDR: 0x00000000  LBEG    : 0x4000c2e0  LEND    : 0x4000c2f6  LCOUNT  : 0x00000000  
      
      ELF file SHA256: 0000000000000000
      
      Backtrace: 0x401420c8:0x3ffb1ef0 0x401420f9:0x3ffb1f10 0x400d1b86:0x3ffb1f30 0x400d1d1b:0x3ffb1f50 0x400d1d92:0x3ffb1f90 0x400d3dd5:0x3ffb1fb0 0x400899b2:0x3ffb1fd0
      
      Rebooting...

**Comment by pschatzmann** [comment, marked as answer]
You will need to analyze the stacktrace to figure out what the issue is.

My gut feeling is that you just run out of memory. Audio is using a lot of memory so you should use either PSRAM or File to store the audio. 

Or even better do not store the audio at all!

ps. it is an interesting idea that the availableForWrite should check the available memory: I think I will extend the logic...

**Reply by pschatzmann** [reply]
Just double checked the code: you need to call ms.begin() to allocate the memory...
